src/plugins/audioplayers/player.c

```diff
@@ -0,0 +1,456 @@
+        // clang-format off
+        platch_call_std(
+            self->channel,
+            "audio.onSeekComplete",
+            &STDMAP2(
+                STDSTRING("player_id"),
+                STDSTRING(self->player_id),
+                STDSTRING("value"),
+                STDBOOL(true)
+            ),
+            NULL,
+            NULL
+        );
+        // clang-format on
+    }
+}
+void audio_player_on_playback_ended(struct audio_player *self) {
+    audio_player_set_position(self, 0);
+    if (audio_player_get_looping(self)) {
+        audio_player_play(self);
+    }
+    if (self->channel) {
+        // clang-format off
+        platch_call_std(
+            self->channel,
+            "audio.onComplete",
+            &STDMAP2(
+                STDSTRING("player_id"),
+                STDSTRING(self->player_id),
+                STDSTRING("value"),
+                STDBOOL(true)
+            ),
+            NULL,
+            NULL
+        );
+        // clang-format on
+    }
+}
+
+void audio_player_set_looping(struct audio_player *self, bool is_looping) {
+    self->is_looping = is_looping;
+}
+
+bool audio_player_get_looping(struct audio_player *self) {
+    return self->is_looping;
+}
+
+void audio_player_play(struct audio_player *self) {
+    if (!self->is_initialized) {
+        return;
+    }
+    audio_player_set_position(self, 0);
+    audio_player_resume(self);
+}
+
+void audio_player_pause(struct audio_player *self) {
+    GstStateChangeReturn ret = gst_element_set_state(self->playbin, GST_STATE_PAUSED);
+    if (ret == GST_STATE_CHANGE_FAILURE) {
+        LOG_ERROR("Unable to set the pipeline to the paused state.\n");
+        return;
+    }
+    audio_player_on_position_update(self);  // Update to exact position when pausing
+}
+
+void audio_player_resume(struct audio_player *self) {
+    if (!self->is_initialized) {
+        return;
+    }
+    GstStateChangeReturn ret = gst_element_set_state(self->playbin, GST_STATE_PLAYING);
+    if (ret == GST_STATE_CHANGE_FAILURE) {
+        LOG_ERROR("Unable to set the pipeline to the playing state.\n");
+        return;
+    }
+    // Update duration when start playing, as no event is emitted elsewhere
+    audio_player_on_duration_update(self);
+}
+
+void audio_player_destroy(struct audio_player *self) {
+    if (self->is_initialized) {
+        audio_player_pause(self);
+    }
+    gst_object_unref(self->bus);
+    self->bus = NULL;
+    gst_object_unref(self->source);
+    self->source = NULL;
+
+    gst_element_set_state(self->playbin, GST_STATE_NULL);
+    gst_object_unref(self->playbin);
+    self->playbin = NULL;
+
+    self->is_initialized = false;
+
+    if (self->url != NULL) {
+        free(self->url);
+        self->url = NULL;
+    }
+
+    if (self->player_id != NULL) {
+        free(self->player_id);
+        self->player_id = NULL;
+    }
+
+    if (self->channel != NULL) {
+        free(self->channel);
+        self->channel = NULL;
+    }
```

Comment:
Hmm this is my mistake.
Right now it is only used in `deinit` function and I expected for caller to free pointer, after it is destroyed.
So maybe it would be better to free by `destroy` function

Reply:
Generally, I think if your API (so `audio_player_new`) handles the allocation, it should also do the freeing. I can't think of a scenario where you'd want to keep the pointer around for longer than the destroy. Noone else besides the `player.c` file knows what struct is behind that pointer anyway (since the actual declaration of `struct audio_player` is only visible in this file)